### ISSUE TITLE
[x86 gcc] Missing dx side effect in AO_compare_double_and_swap_double_full 

### ISSUE THREAD

**Issue by recoules** [ivmai/libatomic_ops#41]
https://github.com/ivmai/libatomic_ops/blob/ae996e93f0ddfb8e2a93d49c6d7edc1478ad60c1/src/atomic_ops/sysdeps/gcc/x86.h#L550-L562

According to https://www.felixcloutier.com/x86/cmpxchg8b:cmpxchg16b, both CMPXCHG8B and CMPXCHG16B may clobber ax and dx if the comparison between old values and memory fails.

**Problem.** Actually, the compiler is unaware that the value of ```old_val2``` may change between subsequent calls. This issue affects both 32- and 64-bit versions.

For instance, consider the following code:
```c
int main (int argc, char *argv[])
{
  AO_double_t v = 0;
  while (!AO_compare_double_and_swap_double_full(&v,0,1,1,1));
  /* not reachable */
  printf("%#016llx\n", v);
  return 0;
}
```
(1) it did not infinitely loop as expected (v is not equal to 0x0000000100000000);
(2) due to optimization, it prints 0 where it was at least expected to print 0x0000000100000001.

**Solution.** An output tied to dx should be added to these blocs:
```c
__asm__ __volatile__("lock; cmpxchg16b %0; setz %1"
                                    : "=m"(*addr), "=a"(result), "=d" (old_val2)
                                    : "m"(*addr), "d" (old_val2), "a" (old_val1),
                                      "c" (new_val2), "b" (new_val1)
                                    : "memory");
```

**Note.** It is not the purpose of this issue, but it is not optimal to always return the result in ax. I can see different solutions if you are interested.

**Comment by ivmai** [comment]
OK, I will fix it.

>  It is not optimal to always return the result in ax. I can see different solutions if you are interested

Yes, I'm ready accept improvement patches.

Thank you!

**Comment by recoules** [comment]
> Yes, I'm ready accept improvement patches.

Yet it is simply small improvements, but:
- I would suggest to give only one input-output ```"+"``` entry for ```*addr``` instead of two entries
- If the compiler support flag outputs, then we can get rid of one instruction
```c
#ifdef  __GCC_ASM_FLAG_OUTPUTS__
__asm__ __volatile__("lock cmpxchg16b %0"
                                    : "+m" (*addr), "=@ccz" (result),
                                      "+d" (old_val2), "+a" (old_val1) 
                                    : "c" (new_val2), "b" (new_val1)
                                    : "memory");
#endif
```
- if not, it is still possible to let the compiler choose the register of its choice using multi-alternative (but I agree it is ugly and may cause the compiler to reserve 2 additional cells in the stack in the caller preamble)
```c
AO_t scratch1, scratch2;
__asm__ __volatile__("lock cmpxchg16b %0" "\n\t" "setz %1"
                                    : "+m,m,m" (*addr), "=a,d,g" (result),
                                      "=d,g,d" (scratch2), "=g,a,a" (scratch1)
                                    : "d,d,d" (old_val2), "a,a,a" (old_val1),
                                      "c,c,c" (new_val2), "b,b,b" (new_val1)
                                    : "memory");
```

**Comment by recoules** [comment]
No problem, I think #42 is safe, but #43 may need some reviews.

However, I don't understand why the patches fail to pass Travis tests. In fact, I looked over the logs and, correct me, but both fail on code unrelated to the changes. Is the master branch currently passing all the tests?

**Comment by ivmai** [comment]
> I don't understand why the patches fail to pass Travis tests.

Yes, master branch has cppcheck fails too, most probably not a bug - I will address it first.

**Comment by ivmai** [comment]
\+ @hboehm

**Comment by ivmai** [comment]
> both CMPXCHG8B and CMPXCHG16B may clobber ... dx

@recoules , good catch!!! The bug existed for 15 years in libatomic_ops! Thank you!

**Comment by ivmai** [comment]
I've started to process your patches.

**Comment by ivmai** [comment]
The bug fix is committed as d728ce4e

**Comment by ivmai** [comment]
Cherry-picked to release branches:
* 7.2 - 31f303835628
* 7.4 - 43addfe1670e
* 7.6 - 50b8b0b7347
